apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,357 @@
+using Grpc.Auth;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// Base class for API-specific builders.
+    /// </summary>
+    /// <typeparam name="TBuilder">The type of this builder, for chaining purposes</typeparam>
+    /// <typeparam name="TClient">The type of client created by this builder.</typeparam>
+    public abstract class ClientBuilder<TBuilder, TClient> where TBuilder : ClientBuilder<TBuilder, TClient>
+    {
+        /// <summary>
+        /// The endpoint to connect to, or null to use the default endpoint.
+        /// </summary>
+        public ServiceEndpoint Endpoint { get; private set; }
+
+        /// <summary>
+        /// The scopes to use, or null to use the default scopes.
+        /// </summary>
+        public IReadOnlyList<string> Scopes { get; private set; }
+
+        /// <summary>
+        /// The channel credentials to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public ChannelCredentials ChannelCredentials { get; private set; }
+
+        /// <summary>
+        /// The path to the credentials file to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public string CredentialsPath { get; private set; }
+
+        /// <summary>
+        /// The credentials to use as JSON, or null if credentials are being provided in a different way.
+        /// </summary>
+        public string JsonCredentials { get; private set; }
+
+        /// <summary>
+        /// The token access method to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public Func<string, CancellationToken, Task<string>> TokenAccessMethod { get; private set; }
+
+        /// <summary>
+        /// The call invoker to use, or null to create the call invoker when the client is built.
+        /// </summary>
+        public CallInvoker CallInvoker { get; private set; }
+
+        /// <summary>
+        /// Creates a new instance with no settings.
+        /// </summary>
+        protected ClientBuilder()
+        {
+        }
+        
+        /// <summary>
+        /// Creates a clone of the given builder, with respect to settings in this class.
+        /// </summary>
+        /// <param name="builder">The builder to clone</param>
+        protected ClientBuilder(TBuilder builder)
+        {
+            Endpoint = builder.Endpoint;
+            Scopes = builder.Scopes;
+            ChannelCredentials = builder.ChannelCredentials;
+            CredentialsPath = builder.CredentialsPath;
+            JsonCredentials = builder.JsonCredentials;
+            TokenAccessMethod = builder.TokenAccessMethod;
+            CallInvoker = builder.CallInvoker;
+        }
+
+        // TODO: WithEndpoint(string) and WithEndpoint(string, int)?
```

Comment:
If ServiceEndpoint already has constructors or Create methods with these overloads, maybe you can skip this TODO or leave for way into the future.

Reply:
ServiceEndpoint has a (string, int) constructor. It should probably have a Parse(string) method as well. But even then, I think it would be good to provide the methods for the sake of convenience. It's very likely that if someone *does* have an endpoint, they've either got a string or a string and integer from a configuration file.

Reply:
OK